text/0000-confidential-token-ids.md

```diff
@@ -0,0 +1,270 @@
+
+To derive a sorting order without revealing the transaction type, the `fee` in the WellFormedTxContext becomes a `fee_ratio`, which is calculated based on the ratio between the minimum fees configured by the node operator on startup. For example, the network operators may have configured the following fees:
+
+```
+{
+    asset_000_fee = 1,
+    asset_111_fee = 2,
+    asset_888_fee = 3,
+}
+```
+
+The ratio between the fees for the `token_ids` is the following, using the least common multiple of the fees:
+
+`token_id` | fee | `fee_ratio` multiplier
+--- | --- | ---
+000 | 1 | 6
+111 | 2 | 3
+888 | 3 | 2
+
+
+Then, imagine that a block contains the following set of transactions with the given fees:
+
+```
+[
+    transaction_1 {token_id: 000, fee: 1},
+    transaction_2 {token_id: 000, fee: 1.5},
+    transaction_3 {token_id: 111, fee: 2},
+    transaction_4 {token_id: 888, fee: 3},
+    transaction_5 {token_id: 888, fee: 4}
+]
+```
+
+The sort order would be:
+
+`transaction_id` | `token_id` | `fee_ratio` multiplier | original fee | `fee_ratio` result
+--- | --- | --- | --- | ---
+transaction_2 | 000 | 6 | 1.5 | 9
+transaction_5 | 888 | 2 | 4 | 8
+transaction_1 | 000 | 6 | 1 | 6
+transaction_3 | 111 | 3 | 2 | 6
+transaction_4 | 888 | 2 | 3 | 6
+
+Note how the two transactions paying more than their minimum fee are sorted to the top, in proportion to how much they are paying above the minimum fee, while all minimum fee transactions have an equal chance of beeing sorted because their `fee_ratio` is equal.
+
+### Performance Impact: Size & Space Analysis
+
+- Each TXO has 6 new bytes allocated in the Amount object for the `masked_token_id` (4 bytes for the value, 2 bytes for Protobuf overhead)
+- The RingCT bulletproof signature increases by at least 96 bytes to contain the Proof of Opening (plus Protobuf overhead).
+- Each block now contains a number of fee outputs scaling linearly with the number of transactions in a block, as opposed to a single fee output aggregated from multiple transactions in a block.
```

Comment:
This is only assuming we stop aggregating the fee outputs (which I would try to push against due to the bloating of the ledger). There was an alternative proposal - I unfortunately don't remember the specifics, but it was something along the lines of number of fee outputs = min(max token_id supported, number of transactions in block) (so there might be some 0 amount outputs).